Generate serialization for SecCertificateRef

mscottapple opened the pull request:
#### da1fed55e56647e8bf8f8d8cdbe3becc16d2c02e
<pre>
Generate serialization for SecCertificateRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=268108">https://bugs.webkit.org/show_bug.cgi?id=268108</a>
<a href="https://rdar.apple.com/121624808">rdar://121624808</a>

Reviewed by beidson (Brady Eidson), sheeparegreat (Simon Lewis), achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder&lt;SecCertificateRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;SecCertificateRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCSecCertificate.h: Added.
(WebKit::CoreIPCSecCertificate::CoreIPCSecCertificate):
(WebKit::CoreIPCSecCertificate::createSecCertificate):
(WebKit::CoreIPCSecCertificate::base64StringFromCertificate):
* Source/WebKit/Shared/cf/CoreIPCSecCertificate.serialization.in: Copied from Source/WebKit/Shared/cf/CFTypes.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/273866@main">https://commits.webkit.org/273866@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92e3f1052ffc22ce970ea4bc6ed5100af60fe0aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38687 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32350 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11944 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31089 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12647 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11063 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11100 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39935 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32713 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32796 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37029 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11248 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9164 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35110 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12993 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8363 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11758 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12079 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mscottapple commented on Source/WebKit/Shared/cf/CFTypes.serialization.in:
```diff
@@ -40,6 +40,9 @@
 [WebKitPlatform, CustomHeader, AdditionalEncoder=StreamConnectionEncoder, ToCFMethod=result->createCFURL()] CFURLRef wrapped by WebKit::CoreIPCCFURL {
 }
 
+[WebKitPlatform, CustomHeader, AdditionalEncoder=StreamConnectionEncoder, ToCFMethod=result->createSecCertificate()] SecCertificateRef wrapped by WebKit::CoreIPCSecCertificate {
```
I don't know what `AdditionalEncoder=StreamConnectionEncoder` does, but all the other CFTypes had this so I kept it for `SecCertificateRef`.

achristensen07 replied:
If you remove it and it won't compile, leave it.
It's needed for types that are sent to the GPU process using StreamConnectionEncoder instead of IPC::Encoder.

sheeparegreat replied:
It's needed if the object is serialized to GPUP.

mscottapple replied:
Yup, doesn't compile if I remove that, so I'll leave it in.

mscottapple commented on Source/WebKit/Shared/cf/CoreIPCSecCertificate.serialization.in:
```diff
@@ -0,0 +1,31 @@
+
+webkit_platform_headers: "CoreIPCSecCertificate.h"
+
+[WebKitPlatform, AdditionalEncoder=StreamConnectionEncoder] class WebKit::CoreIPCSecCertificate {
```
I don't know what `AdditionalEncoder=StreamConnectionEncoder` does, but all the other CFTypes had this so I kept it for `SecCertificateRef`.

mscottapple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -19006,7 +19014,6 @@
 			isa = PBXTargetDependency;
 			platformFilters = (
 				ios,
-				xros,
```
I don't know why these were removed. Do I need to put them back?

achristensen07 replied:
I would.  This change is probably automatically done by Xcode, but it's not relevant to the change you made and could make reverts messier than necessary.

beidson replied:
Yes. Xcode fights about this when you don't have the Vision Pro SDK installed, but they need to be put back in.

sheeparegreat replied:
Yes. Don't delete those

mscottapple commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -25,6 +25,10 @@
 
 #pragma once
 
+#if USE(CF)
```
I'm pretty sure nobody is going to like this, but I couldn't find another place to put this that worked for all parent `#include`s. It really should be in `"GeneratedSerializers.h"` but I don't know how to make that happen.

achristensen07 replied:
Yeah, this isn't good.  We also don't want it in GeneratedSerializers.h.  All we want in GeneratedSerializers.h is "typedef struct __SecTrust *SecTrustRef;"

sheeparegreat replied:
`headers: <Security/Security.h>` in the serialization.in should help?

mscottapple replied:
Unfortunately, it does not.

mscottapple replied:
On the plus side, the next 3 types I plan to tackle are the `Sec*Ref` types, which all will fall under this one header.

achristensen07 replied:
We don't want to include Security.h here.  That is too broad, which slows down build times, and "typedef struct __SecTrust *SecTrustRef;" will do just fine.

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCSecCertificate.h:
```diff
@@ -0,0 +1,80 @@
+    {
+        RELEASE_ASSERT(certificate);
+        NSData *certificateData = (__bridge NSData*)SecCertificateCopyData(certificate);
+        NSString *base64String = [certificateData base64EncodedStringWithOptions:0];
```
Let's just encode the NSData instead of adding a base64 roundtrip

sheeparegreat commented on Source/WebKit/Shared/cf/CoreIPCSecCertificate.serialization.in:
```diff
@@ -0,0 +1,31 @@
+webkit_platform_headers: "CoreIPCSecCertificate.h"
+
+[WebKitPlatform, AdditionalEncoder=StreamConnectionEncoder] class WebKit::CoreIPCSecCertificate {
+    String m_certificateData;
```
What validation can be added to this?

sheeparegreat replied:
Can this be broken down further?

sheeparegreat commented on Source/WebKit/Shared/cf/CoreIPCSecCertificate.h:
```diff
@@ -0,0 +1,80 @@
+        NSData *data = [[NSData alloc] initWithBase64EncodedString:(NSString *)m_certificateData options:0];
+        RELEASE_ASSERT(data);
+
+        auto result = adoptCF(SecCertificateCreateWithData(kCFAllocatorDefault, bridge_cast(data)));
```
Can you break SecCertificateRef down into smaller pieces? This is passing unvalidated data into the constructor.

mscottapple replied:
Unfortunately, a `SecCertificateRef` can only be created with `SecCertificateCreateWithData()`.

sheeparegreat reviewed (changes requested):
Can the object be broken down further into smaller pieces which can be validated?

achristensen07 commented on Source/WebKit/Shared/cf/CoreIPCSecCertificate.h:
```diff
@@ -0,0 +1,81 @@
+    RetainPtr<SecCertificateRef> createSecCertificate() const
+    {
+        auto certificate = adoptCF(SecCertificateCreateWithData(kCFAllocatorDefault, m_certificateData.get()));
+        RELEASE_ASSERT(certificate);
```
I don't see any reason to add any of these RELEASE_ASSERTS.  Transformation failures result in null data/cert.  That is existing behavior.  Let's not add a crash with this PR.

mscottapple committed:
Generate serialization for SecCertificateRef
https://bugs.webkit.org/show_bug.cgi?id=268108
rdar://121624808

Reviewed by beidson (Brady Eidson), sheeparegreat (Simon Lewis), achristensen07 (Alex Christensen).

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::ArgumentCoder<SecCertificateRef>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<SecCertificateRef>>::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCSecCertificate.h: Added.
(WebKit::CoreIPCSecCertificate::CoreIPCSecCertificate):
(WebKit::CoreIPCSecCertificate::createSecCertificate):
(WebKit::CoreIPCSecCertificate::base64StringFromCertificate):
* Source/WebKit/Shared/cf/CoreIPCSecCertificate.serialization.in: Copied from Source/WebKit/Shared/cf/CFTypes.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/273866@main
